K claire.pp

hano-ios opened the pull request:
# 성적관리 [STEP 1] claire.pp
@zdodev 
STEP 1 PR입니다.

안녕하세요 claire입니다
제가 과제하면서 고민했던 것은
- 하나의 메서드가 하나의 기능만 수행하도록 작성하기
- 적절한 단위로 끊어서 커밋하기
- 커밋 로그 잘 작성해보기

이었습니다.

## 하나의 메서드가 하나의 기능만 수행하도록 작성하기
중요한 원칙이기도 하고, 한 메서드당 10줄 이내라는 제약이 있기 때문에 고민을 했습니다. 
이러한 부분에서 가장 의문이 생겼던 것은 run 메서드였습니다.

### run()
run() 함수는 메뉴를 관리하는 MenuManager 클래스의 핵심 메서드입니다.
처음에 제가 계획했던 run() 함수의 기능은 다음과 같았습니다.
- 메뉴 출력
- 메뉴에 대한 입력받기
- 각 메뉴에 해당하는 기능을 수행
- 계속 반복할지 아닐지 결정

이렇게 했더니 당연히 10줄이 넘어갔고 다음과 같이 분리했습니다

run()의 기능
- 메뉴 출력
- 메뉴에 대한 입력을 받기
- 계속 반복할지 아닐지 결정

inputToMenu()의 기능
- 입력에 대한 기능을 수행

여기서 제가 가진 의문은
- run()이 사실상 세 가지 기능을 수행하고 있는데, 잘못된 것인지?
- inputToMenu가 각 케이스에 대해 한줄씩만 적어도 10줄이 넘어가는데 더 줄일 수 있는 방법이 있는지?
였습니다.

add의 경우에는 
- 매개변수가 nil인지 판별
- 이미 해당 학생이 존재하는지 판별
- 학생 정보를 추가

이렇게 세 가지 기능을 하지만, 결국 이것이 학생을 정보를 추가하는 한 가지 기능이라고 생각했습니다.
run도 마찬가지로 메뉴가 동작하게하는 한가지 메서드라고 생각하면 괜찮은 것 같은데.. 헷갈려서 질문을 드리고 싶습니다!

또, inputToMenu의 코드를 줄이는 방법에 대해서
1. 열거형 Command 내부에 각 케이스별로 수행할 기능을 정의한 함수를 만든다
> 예시)
enum Command {
...
func menuFunction() {
switch self {
case .studentAdd:
return add()
...

2. inputToMenu에서 Command.allCases 중 현재 input과 일치하는 것이 있다면, 해당 함수를 실행하도록 한다

라고 생각했지만, Command 내부에는 10줄 이상의 함수가 생기고(switch문은 동일..)
2번의 과정에서 forEach 내부에 if절이 생겨버려 2번의 들여쓰기 제한을 넘기게 되었습니다
그래서 이건 어쩔 수 없이 10줄을 넘겨야만 하는 것인지, 다른 방법이 있는 것인지 궁금합니다!

## 커밋
커밋 단위를 기능단위로 하자!는 Ground rule을 만들고, 최대한 기능 단위별로 커밋하려고 했습니다.
또, 커밋 로그는 알아보기 쉬운 한글로 작성하고자 했고, subject의 첫 단어를 현재형 동사를 사용하는 것 같아
[추가],[수정],[제거] 와 같은 형태로 새로운 규칙을 만들어 커밋했습니다.
커밋 단위를 이렇게 잡는 게 맞는건지 (함수 하나하나..) 또 로그 작성은 이렇게 하는 것이 맞는지 궁금합니다!

## 마무리
여기까지가 제 과제에 대한 설명과, 제 의문점들에 대한 내용이었습니다.
긴 글 읽어주셔서 감사드리며,
제 의문점들과 기타 제가 놓친 점들에 대한 피드백 해주시면 정말 감사하겠습니다!

Commit:
feat:  [추가] 입출력, 메뉴 관리 프로토콜, 학생 구조체

프로그램의 뼈대를 구성하기 위해 입출력 프로토콜 InputManagable과
메뉴 관리 프로토콜인 MenuManagable
그리고 학생에 대한 정보를 구성할 Student 구조체를 추가했습니다.

Commit:
feat: InputManager 클래스 추가, readUserInput, toStudent 구현

사용자 입력을 관리하는 클래스인 InputManager를 추가함
InputManagable 프로토콜을 채택하므로,
입력을 받는 readUserInput 메서드와
입력을 통해 Student 객체를 생성하는 toStudent 메서드를 구현함

Commit:
feat: MenuManager 클래스 추가

MenuManagable 프로토콜을 채택하는 MenuManager 클래스 추가
성적 관리 프로그램의 메뉴를 관리하는 역할
내부 변수 students, inputManager만 선언

Commit:
feat: [수정] MenuManagable의 students를 Set으로 수정, Student에 Hasable 속성 추가

MenuManagable의 students에 들어가는 Student 객체의 이름에 중복이 없도록 하기 위해
students를 Set으로 수정하였고, 이를 위해 Student에 Hashable 속성을 추가했음

Commit:
feat: [추가] MenuManager의 add 메서드 구현

학생 정보를 매개변수로 받아 students에 추가해주는 add 메서드를 구현함

Commit:
feat: [수정] MenuManagable의 find 메서드의 반환값을 Student?로 수정, 구현

MenuManagable에서 students는 학생들에 대한 정보가 담겨 있음
find는 students안에 특정 이름을 가진 학생이 있는지를 찾아내는 메서드
추후 삭제나 수정에서 반환값이 Student인 경우가 더 편할 것이라 생각하여
find의 반환값을 Student로 수정하고 구현함

Commit:
feat: [추가] MenuManager의 delete 메서드 구현

학생 정보에서 특정 학생을 삭제하는 메서드인 delete를 구현함

Commit:
chore: guard 구문 수정

else 사용 불가이므로 guard-else도 사용 불가인 줄 알았으나
guard 사용 가능하므로 해당하는 구문들 수정함

Commit:
feat: [추가] Command, Behavior 열거형 추가

MenuManager에서 사용자의 입력을 정의한 열거형 Command와
출력으로 띄울 메시지들을 정의한 Behavior를 추가함

Commit:
feat: [추가] InputManager.toMenu() 추가

사용자의 메뉴 입력을 받고 검증하는 메서드 toMenu를 추가함

Commit:
chore: [수정] 들여쓰기 수정

가독성을 위해 들여쓰기를 수정함

Commit:
feat: [수정] toStudent, Command, Behavior, add, delete, run

toStudent 함수에 메시지를 매개변수로 받아 출력하도록 수정
Command, Behavior의 값 수정
add, delete에 존재하는 학생에 따른 동작 추가, 옵셔널 매개변수
메뉴를 계속해서 출력해주고 입력을 받는 run 함수 추가
run함수에 입력에 따른 메뉴를 연결하는 부분을 inputToMenu로 분리

Commit:
feat: [추가] 메인문

구현된 클래스를 바탕으로 메인문 작성

Commit:
chore: [수정] add, delete

혹여 insert, remove단계에서 오류가 발생할 수 있으므로
해당 메시지는 해당 명령 수행 후 출력하도록 함
중요도가 높지 않아 chore로 commit

Commit:
feat: [추가] 3~5번 메뉴 출력

아직 준비하지 않은 기능이라는 것만 출력하도록 함

Commit:
chore: [제거] find 메서드 삭제

사용하지 않는 find 메서드 삭제
contains 메서드로 대체 가능

Commit:
chore: [수정] delete 순서 수정

가독성을 위한 코드 수정

Commit:
chore: [수정] run 코드 수정

코드를 더 간단하게 줄였음

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
-
 import Foundation
+
+protocol InputManagable {
+    func readUserInput() -> String?
+    func toStudent(message: String) -> Student?
+    func toMenu() -> String?
+}
```
`InputManagable` 프로토콜을 정의해보셨네요. 좋습니다!

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+    func toMenu() -> String?
+}
+
+protocol MenuManagable {
+    var students: Set<Student> { get set }
+    func run() -> Bool
+    func add(student: Student?)
+    func delete(student: Student?)
+}
```
`MenuManagable` 프로토콜을 정의해보셨네요. 좋습니다!

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+    func delete(student: Student?)
+}
+
+struct Student {
+    let name: String
+}
```
학생에 대한 모델을 구현하셨네요. 좋습니다!

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+}
+
+class InputManager: InputManagable {
+    func readUserInput() -> String? {
+        guard let userInput: String = readLine(), userInput != "" else { return nil }
+        return userInput
+    }
```
`readUserInput` 메서드의 목적이 입력 값이 공백인지 아닌지 판단하기 위한 메서드인지 궁금합니다!🤔

hano-ios replied:
사용자 입력을 받을 때 항상 공백인지 아닌지 판별해야 하기 때문에 해당부분을 readUserInput으로 구현해서 readLine 대신 사용하도록 했습니다!

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+
+class MenuManager: MenuManagable {
+    enum Command: String, CaseIterable {
+        case studentAdd = "1", studentDelete = "2", gradeAdd = "3", gradeDelete = "4", show = "5", end = "X"
```
너무 길어질 경우 enum 케이스를 가로로 쭉 나열하는 것 보다는 가독성 측면에서 적절하게 케이스 별로 개행을 하는 것은 어떨까요? 의견이 궁금합니다!🤔

hano-ios replied:
아래 enum과 마찬가지로 수정해보겠습니다!

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+    }
+    
+    enum Behavior {
+        case menu, wrongMenu, needAddName, needDeleteName, wrongInput, nameExist(Student), noStudent(Student), addedStudent(Student), deletedStudent(Student), programEnd
```
마찬가지로 enum 케이스를 가로로 쭉 나열하는 것 보다는 가독성 측면에서 적절하게 케이스 별로 개행을 하는 것은 어떨까요? 의견이 궁금합니다!🤔

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+    enum Behavior {
+        case menu, wrongMenu, needAddName, needDeleteName, wrongInput, nameExist(Student), noStudent(Student), addedStudent(Student), deletedStudent(Student), programEnd
+        
+        var output: String {
```
출력 문자열을 모아서 정의해주셨네요 좋습니다! 용도에 따라 나눠서 구현해도 좋을 것 같습니다.😁

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+    }
+    
+    func inputToMenu(input: String) {
+        switch input {
+        case Command.studentAdd.rawValue:
+            add(student: inputManager.toStudent(message: Behavior.needAddName.output))
+        case Command.studentDelete.rawValue:
+            delete(student: inputManager.toStudent(message: Behavior.needDeleteName.output))
+        case Command.end.rawValue:
+            print(Behavior.programEnd.output)
+        case Command.gradeAdd.rawValue, Command.gradeDelete.rawValue, Command.show.rawValue:
+            print("아직 준비되지 않은 기능입니다.")
+        default:
+            print(Behavior.wrongMenu.output)
+        }
```
적은 케이스가 있을 때는 현재 구문도 좋은 구문이라고 생각합니다. 다만 이후에 많은 케이스가 생길 수 있는 부분은 추상화와 다형성을 통해서 접근해 볼 수 있지 않을까 싶습니다. 지금 이런 확장에 대한 고민이 생기신 것만으로도 좋은 발전이라고 생각합니다.🤗

zdodev commented on GradeManager/main.swift:
```diff
@@ -1,7 +1,142 @@
+        self.students = Set<Student>()
+    }
+
+    func run() -> Bool {
+        print(Behavior.menu.output)
+        guard let userInput = inputManager.toMenu() else { print(Behavior.wrongMenu.output); return true }
+        inputToMenu(input: userInput)
+        return userInput != Command.end.rawValue
+    }
```
`run` 메서드를 잘못 구현했다고 생각되지는 않습니다. `add` 메서드처럼 기능을 종합해서 하나의 기능으로 구현한 단위로 보여집니다. 그리고 메서드의 역할을 분리하여 `run` 메서드에서 적절하게 사용했다는 생각이 들었습니다.😁

zdodev reviewed (changes requested):
안녕하세요. @HanEun-ji !
첫 번째 프로젝트, 첫 번째 PR 축하합니다~

메서드 역할 분리를 적절하게 잘 작성해주신 점이 인상 깊었습니다!
프로토콜과 옵셔널, 네이밍에도 신경쓰신 점이 눈에 띄었습니다.

다만 모든 코드를 한 파일에 작성하셨는데, 타입 별로 파일을 적절하게 분리해보면 좋을 것 같습니다.😁

남겨주신 질문에 대해서 다양한 의견을 남겨보았습니다.
궁금하신 점이 있다면 같이 의견을 나눠보면 좋겠습니다.🤗

Commit:
chore: [수정] 개행, 열거형 Behavior와 Error 분리

가독성을 위해 몇가지 구문을 개행했고,
열거형 Behavior를 목적에 따라 안내 메시지 Behavior와
에러 메시지 출력을 위해 필요한 Error로 분리함

Commit:
chore: [분리] 파일 클래스별로 분리

하나의 파일로 뭉쳐있던 main.swift를 4개로 분리했음

hano-ios commented:
안녕하세요! 이렇게 상세하게 리뷰해주셔서 정말 감사합니다 😮
남겨주신 코멘트대로
- 가독성 수정 (개행)
- 열거형 목적에 따라 분리(Behavior, Error)
- 파일 클래스별로 분리
해보았습니다. 한 번 검토해주시면 감사하겠습니다!

zdodev commented on GradeManager/MenuManager.swift:
```diff
@@ -0,0 +1,116 @@
+}
+
+class MenuManager: MenuManagable {
+    enum Command: String, CaseIterable {
+        case studentAdd = "1", studentDelete = "2",
+             gradeAdd = "3", gradeDelete = "4",
+             show = "5", end = "X"
```
보다 읽기 수월해졌습니다.👍

zdodev commented on GradeManager/MenuManager.swift:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+    
+    enum Behavior {
```
enum의 역할을 잘 나눠주셨네요.👍

zdodev commented on GradeManager/MenuManager.swift:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+    
+    enum Error {
```
enum의 역할을 잘 나눠주셨네요.👍

zdodev reviewed (approved):
안녕하세요. @HanEun-ji !

enum 케이스 개행 및 역할 분리로 인해 가독성이 더 좋아졌습니다!👍
그리고 적절한 파일 분할도 잘해주셨습니다!😁

스텝 1 진행하시면서 작성하신 커밋 단위도 좋았습니다.
그리고 커밋 제목을 일관성있고 어떤 작업을 수행했는지에 대해서 잘성해주셨습니다!
커밋 단위 및 내용에 대해서는 절대적인 규칙은 없지만 본인만의 기준을 일관성있게 만들어 나가는 것도 좋을 것 같습니다!😁

스텝 1 구현하시느라 고생하셨습니다. 이번 스텝은 여기서 마무리 하겠습니다.
다음 스텝도 화이팅입니다!🤗
